plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts

```diff
@@ -0,0 +1,268 @@
+      .where('status', 'processing')
+      .andWhere(
+        'last_heartbeat_at',
+        '<=',
+        this.db.client.config.client === 'sqlite3'
+          ? this.db.raw(`datetime('now', '-${Number(timeoutS)} seconds')`)
+          : this.db.raw(`dateadd('second', -${Number(timeoutS)}, now())`),
+      );
+    const tasks = rawRows.map(row => ({
+      taskId: row.id,
+    }));
+    return { tasks };
+  }
+
+  async completeTask({
+    taskId,
+    status,
+    eventBody,
+  }: {
+    taskId: string;
+    status: Status;
+    eventBody: JsonObject;
+  }): Promise<void> {
+    let oldStatus: string;
+    if (status === 'failed' || status === 'completed') {
+      oldStatus = 'processing';
+    } else {
+      throw new Error(
+        `Invalid status update of run '${taskId}' to status '${status}'`,
+      );
+    }
+    await this.db.transaction(async tx => {
+      const [task] = await tx<RawDbTaskRow>('tasks')
+        .where({
+          id: taskId,
+        })
+        .limit(1)
+        .select();
+
+      if (!task) {
+        throw new Error(`No task with taskId ${taskId} found`);
+      }
+      if (task.status !== oldStatus) {
+        throw new ConflictError(
+          `Refusing to update status of run '${taskId}' to status '${status}' ` +
+            `as it is currently '${task.status}', expected '${oldStatus}'`,
+        );
+      }
+      const updateCount = await tx<RawDbTaskRow>('tasks')
+        .where({
+          id: taskId,
+          status: oldStatus,
+        })
+        .update({
+          status,
+        });
+      if (updateCount !== 1) {
+        throw new Error(
+          `Failed to update status to '${status}' for taskId ${taskId}`,
+        );
+      }
+
+      await tx<RawDbTaskEventRow>('task_events').insert({
+        task_id: taskId,
+        event_type: 'completion',
+        body: JSON.stringify(eventBody),
+      });
+    });
+  }
+
+  async emitLogEvent({ taskId, body }: TaskStoreEmitOptions): Promise<void> {
+    const serliazedBody = JSON.stringify(body);
+    await this.db<RawDbTaskEventRow>('task_events').insert({
+      task_id: taskId,
+      event_type: 'log',
+      body: serliazedBody,
+    });
+  }
+
+  async listEvents({
+    taskId,
+    after,
+  }: TaskStoreGetEventsOptions): Promise<{ events: DbTaskEventRow[] }> {
+    const rawEvents = await this.db<RawDbTaskEventRow>('task_events')
+      .where({
+        task_id: taskId,
+      })
+      .andWhere(builder => {
+        if (typeof after === 'number') {
+          builder.where('id', '>', after).orWhere('event_type', 'completion');
+        }
+      })
+      .select();
+
+    const events = rawEvents.map(event => {
+      try {
+        const body = JSON.parse(event.body) as JsonObject;
+        return {
+          id: event.id,
+          taskId,
+          body,
+          type: event.event_type,
+          createdAt: event.created_at,
+        };
+      } catch (error) {
+        throw new Error(
```

Comment:
so many json parse guards, really necessary? we serialize ourselves anyway right